kernel/runtime/cxx/runtime/futex.cc

```diff
@@ -167,6 +169,85 @@ static int futex_wake(
     return 0;
 }
 
+static int futex_requeue(uint32_t *uaddr /*&barrier*/, unsigned int flags, uint32_t *uaddr2 /*target*/, uint32_t val/*wake limit*/, uint32_t val2/*requeue limit*/){
+    MLOG_DETAIL(mlog::app, "Futex_requeue", DVARhex(uaddr), DVARhex(uaddr2), DVAR(val), DVAR(val2));
+    uint32_t hashFrom = hash_futex(uaddr);
+
+    // wakes up a maximum of  val  waiters that  are  waiting  on  the  futex at uaddr
+    mythos::Mutex::Lock guard0(requeueLock);
+    mythos::Mutex::Lock guard1(queue[hashFrom].readLock);
+    auto curr = &(queue[hashFrom].queueHead);
+    for (int i = 0; i < val; i++) {
+        while (curr->load(std::memory_order_relaxed) && curr->load(std::memory_order_relaxed)->uaddr != uaddr) {
+            MLOG_DETAIL(mlog::app, "Skip entry", DVARhex(curr->load(std::memory_order_relaxed)->uaddr),
+                        DVAR(curr->load(std::memory_order_relaxed)->ec), DVAR(curr->load(std::memory_order_relaxed)->queueNext), DVARhex(uaddr) );
+            curr = &curr->load(std::memory_order_relaxed)->queueNext;
+        }
+
+        if (curr->load(std::memory_order_relaxed) == nullptr){
+          MLOG_DETAIL(mlog::app, "requeue end (end of queue reached)" );
+          return 0;
+        }
```

Comment:
Comment missing ... this part scans for the next waiter in the queue?

Reply:
added comment